### PR TITLE
Salt orchestrate for (single) node upgrade

### LINKED ISSUE #934: Create a node Upgrade runner
Create a salt runner that do the  workload of control plane node upgrade.

The salt runner should look at value from apiserver (though ext_pillar) to know which version of metalk8s **should** be present on a node.

If this is a newer version that the one already deployed run the highstate on this node to upgrade to the required version.

https://kubernetes.io/docs/tasks/administer-cluster/cluster-management/#maintenance-on-a-node

```
# Using salt external pillar reading the apiserver
Get new_version in node “labels” “metalk8s.scality.com/version”
 
Call cordon node				# Mark node as unschedulable
Call drain node				# gracefully terminate all pods on node
   
Call highstate node			        # Salt highstate

Call uncordon node				# Mark node as schedulable
```
To `drain` #979 is needed

### PR DESCRIPTION
**Pull request by TeddyAndrieux**
```
salt-run state.orchestrate metalk8s.orchestrate.deploy_node saltenv=metalk8s-2.0 pillarenv=metalk8s-2.0 pillar="{'orchestrate': {'node_name': 'bootstrap'}}"
```

Fixes: #934

### CODE REVIEW

File: salt/metalk8s/orchestrate/upgrade/node.sls [hunk 1]
```diff
@@ -0,0 +1,61 @@
+{%- set dest_version = pillar.orchestrate.dest_version %}
+{%- set node_name = pillar.orchestrate.node_name %}
+
+{%- set kubeconfig = "/etc/kubernetes/admin.conf" %}
+{%- set context = "kubernetes-admin@kubernetes" %}
+
+{%- if pillar.metalk8s.nodes[node_name]['version'] == dest_version %}
+
+Node {{ node_name }} already in version {{ dest_version }}:
```
Comment (NicolasT) [inline salt/metalk8s/orchestrate/upgrade/node.sls]: Shouldn't we simply call this `deploy_node` or something, and always run everything?

File: salt/metalk8s/orchestrate/upgrade/node.sls [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+Cordon node {{ node_name }}:
+  metalk8s_kubernetes.node_cordoned:
+    - name: {{ node_name }}
+    - kubeconfig: {{ kubeconfig }}
+    - context: {{ context }}
+
+# TODO: Call drain
+
+Run the highstate on {{ node_name }}:
```
Comment (NicolasT) [inline salt/metalk8s/orchestrate/upgrade/node.sls]: I think we should *first* set the version label, *then* cordon/drain/apply highstate/uncordon, and stop.

Comment (NicolasT) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: In which case the pillar override wouldn't be required.

File: salt/metalk8s/orchestrate/upgrade/node.sls [hunk 3]
```diff
@@ -0,0 +1,61 @@
+{%- set dest_version = pillar.orchestrate.dest_version %}
```
Comment (NicolasT) [inline salt/metalk8s/orchestrate/upgrade/node.sls]: Can't this be basically the same as, or almost the same, as `deploy_new_node.sls`?

Comment (NicolasT) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: I.e. don't have two playbooks.

**Review by NicolasT — CHANGES_REQUESTED** [review]

**Comment by NicolasT** [comment]
What I'm trying to say in the above: do we really need a 'single node upgrade' playbook?

File: salt/metalk8s/orchestrate/upgrade/node.sls (hunk 2 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: Hmm for me we should set the version label only when the node is really running this version so at the end. 
And then we not execute anything on node which are already in the correct version and if we got an error during drain cordon highstate or uncordon we don't change the label version.

File: salt/metalk8s/orchestrate/upgrade/node.sls (hunk 1 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: You mean even for deploy a node we will drain cordon and uncordon ?

File: salt/metalk8s/orchestrate/upgrade/node.sls (hunk 3 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: I create one for single as we will use this for control_plane upgrade and for workload_plane_upgrade 
And then have something like that for upgrade global runner
```
└> cat salt/metalk8s/orchestrate/upgrade/init.sls 
Execute the upgrade prechecks:
  salt.runner:
    - name: state.orchestrate
    - mods:
      - metalk8s.orchestrate.upgrade.precheck

Upgrade etcd cluster:
  salt.runner:
    - name: state.orchestrate
    - mods:
      - metalk8s.orchestrate.upgrade.etcd
    - saltenv: {{ saltenv }}

Upgrade control_plane nodes:
  salt.runner:
    - name: state.orchestrate
    - mods:
      - metalk8s.orchestrate.upgrade.control_plane
    - require:
      - salt: Upgrade etcd cluster

Upgrade workload_plane nodes:
  salt.runner:
    - name: state.orchestrate
    - mods:
      - metalk8s.orchestrate.upgrade.workload_plane
    - require:
      - salt: Upgrade control_plane nodes
```
Not finished not tested of course but that my idea

Comment (NicolasT) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: Hmh, not sure that's the best approach. If we were to do a bit more in this playbook, and basically not use any 'sub'-runners, we can more clearly decide which nodes get upgraded in which stage. As an example, there's no need to bring a control-plane node in highstate as part of the *Upgrade workload-plane nodes*, when they've been put in highstate already.

File: salt/metalk8s/orchestrate/upgrade/node.sls (hunk 1 above)
Comment (NicolasT) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: Yeah, why not? Would basically be a no-op.

File: salt/metalk8s/orchestrate/upgrade/node.sls (hunk 2 above)
Comment (NicolasT) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: No. We're intent-driven, Kubernetes-style. We set the 'desired' version, then we apply actions (as generic as possible) to make that 'desire' reality.

**Comment by NicolasT** [comment]
Maybe we should discuss more, but I'm still not sure we *need* an orchestration like this. Could have a single playbook that does a whole cluster upgrade. Not sure we even want to be able to upgrade a cluster 'partially', which could result in a broken system with incompatible versions.

Except for etcd, maybe.

File: salt/metalk8s/orchestrate/upgrade/node.sls (hunk 2 above)
Comment (NicolasT) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: Could we have a short session today to discuss this and the path forward, @TeddyAndrieux @gdemonet ?

File: salt/metalk8s/orchestrate/upgrade/node.sls (hunk 1 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: Edit: Oh for deploy, hmm yes maybe we can

File: salt/metalk8s/orchestrate/upgrade/node.sls (hunk 2 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: Ok following the discussion we had, lets
- have one orchestrate for all the cluster upgrade
- set the version before upgrading a node
- always do cordon, drain, uncordon

**Comment by TeddyAndrieux** [comment]
@bert-e reset

File: salt/metalk8s/orchestrate/upgrade/node.sls (hunk 2 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/upgrade/node.sls]: This PR is only about one node, I will do the global orchestrate in another PR

File: salt/metalk8s/orchestrate/deploy_node.sls [hunk 4]
```diff
@@ -40,31 +44,46 @@ Refresh the mine:
...
     - require:
       - salt: Set grains
       - salt: Refresh the mine
+      - metalk8s_kubernetes: Cordon the node
+
+Uncordon the node:
+  metalk8s_kubernetes.node_uncordoned:
+    - name: {{ node_name }}
+    - kubeconfig: {{ kubeconfig }}
+    - context: {{ context }}
```
Comment (NicolasT) [inline salt/metalk8s/orchestrate/deploy_node.sls:70]: Wouldn't this `require` `Run the highstate`?

File: ui/src/services/api.js [hunk 5]
```diff
@@ -129,11 +129,10 @@ export async function deployNode(url, token, node) {
       {
         client: 'runner',
         fun: 'state.orchestrate',
-        arg: ['metalk8s.orchestrate.deploy_new_node'],
+        arg: ['metalk8s.orchestrate.deploy_node'],
         kwarg: {
           saltenv: 'metalk8s-2.0',
```
Comment (NicolasT) [inline ui/src/services/api.js:134]: @carlito-scality @MonPote Is there a tracking ticket to get rid of this hard-coded value?

File: salt/metalk8s/orchestrate/deploy_node.sls (hunk 4 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/orchestrate/deploy_node.sls:70]: True, I just forgot this

**Comment by TeddyAndrieux** [comment]
@bert-e reset

**Review by NicolasT — APPROVED** [review]

**Comment by TeddyAndrieux** [comment]
@bert-e approve

File: ui/src/services/api.js (hunk 5 above)
Comment (carlito-scality) [reply ui/src/services/api.js:134]: @NicolasT See https://github.com/scality/metalk8s/issues/1105

**Pull request merged**